tests/goto_wcc_invalid_test.sh

```diff
@@ -0,0 +1,162 @@
+#!/bin/bash
+
+set -o pipefail
+
+source ./test_sub.sh
+
+function test_backward_goto() {
+  begin_test_suite "Backward Goto"
+  
+  compile_error 'simple backward goto' '
+int main() {
+start_label:
+    int x = 1;
+    if (x) {
+        goto start_label;  // BACKWARD GOTO - should fail
+    }
+    return 0;
+}'
+
+  compile_error 'backward goto in nested loops' '
+int main() {
+outer_loop:
+    for (int i = 0; i < 2; i++) {
+        for (int j = 0; j < 2; j++) {
+            if (i == 1 && j == 1) {
+                goto outer_loop;  // BACKWARD GOTO - should fail
+            }
+        }
+    }
+    return 0;
+}'
+
+  compile_error 'backward goto with conditional' '
+int main() {
+retry:
+    int x = 5;
+    if (x > 0) {
+        x--;
+        if (x > 0) {
+            goto retry;  // BACKWARD GOTO - should fail
+        }
+    }
+    return x;
+}'
+
+  end_test_suite
+}
+
+function test_cross_branch_goto() {
+  begin_test_suite "Cross-Branch Goto"
+
+  compile_error 'goto from if to else branch' '
+int main() {
+    int x = 1;
+    if (x == 1) {
+        goto else_label;  // CROSS-BRANCH GOTO - should fail
```

Comment:
It does not determine that you are jumping to another branch, so 
inserting a `if` statement before it will compile successfully and 
return the wrong result.
```c
int main() {
    int x = 1;
    if (x == 1) {
        if (x < 100)
            goto else_label;
    } else {
        if (x == 2)
            return 2;
else_label:
        return 0;
    }
    return 1;
}
```

Reply:
Hm, yes. This requires a new data structure I'm afraid.